src/Clay/Grid.hs

```diff
@@ -1,58 +1,271 @@
+-------------------------------------------------------------------------------
+-- | Property defines the element location inside grid template
+gridArea :: GridArea -> Css
+gridArea = key "grid-area"
+
+blankGridArea :: GridArea
+blankGridArea = GridArea "."
+
+newtype GridArea = GridArea Text
+  deriving (IsString, Val)
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
+gridRowStart = key "grid-row-start"
+
+gridRowEnd :: GridLocation -> Css
+gridRowEnd = key "grid-row-end"
+
+gridColumnStart :: GridLocation -> Css
+gridColumnStart = key "grid-column-start"
+
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocation_Data . GridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
+  | GridLocation_Data GridLocationData
+
+instance Val GridLocation where
+  value (GridLocation_Keyword v) = v
+  value (GridLocation_Data d) = value d
+
+instance Auto    GridLocation where auto    = GridLocation_Keyword auto
+instance Inherit GridLocation where inherit = GridLocation_Keyword inherit
+instance Initial GridLocation where initial = GridLocation_Keyword initial
+instance Unset   GridLocation where unset   = GridLocation_Keyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+data GridLocationData = GridLocationData IsSpan (These Integer GridArea)
+
+instance Val GridLocationData where
+  value (GridLocationData isSpan indexAndOrGridArea) =
+    if isSpan == Span
+    then value (span :: Value, indexAndOrGridArea)
+    else value indexAndOrGridArea
+
+pattern GridIndex :: Integer -> GridLocation
+pattern GridIndex n = GridLocation_Data (GridLocationData NoSpan (This n))
+
+instance Num GridLocation where
+  -- for index literals
+  fromInteger = GridIndex
+  -- for negative index literals
+  negate (GridIndex index) = GridIndex $ negate index
+  -- in general we don't support arithmetic on this type
+  negate _ = error "negate not defined for GridLocation"
+  abs = error "abs not defined for GridLocation"
+  signum = error "abs not defined for GridLocation"
+  (+) = error "addition not defined for GridLocation"
+  (*) = error "multiplication not defined for GridLocation"
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the template for grid layout
+gridTemplateAreas :: GridTemplateAreas -> Css
+gridTemplateAreas = key "grid-template-areas"
+
+newtype GridTemplateAreas = GridTemplateAreas Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+instance IsList GridTemplateAreas where
+  type Item GridTemplateAreas = [GridArea]
+  toList = error "toList GridTemplateAreas is not defined"
```

Comment:
Like for the `Num` instance, you probably just want to be able to write list literals. But that leads to a partial interface, and it is confusing if at some point people try to probe it further and discover this error.

Reply:
Also in this case, can't you just write the definition of `toList`?

Reply:
No, in this case you can't write `toList`, only `fromList`. GridTemplateAreas is defined as a newtype around `Value`. Some but not all values are lists. So you can convert a list to a value, but not vice versa. 

That said, I'm stepping through this and going to remove the partial type class instances. 
I'm going to replace them with pattern synonyms  or smart constructors.